great_ape_safe/ape_api/balancer.py

```diff
@@ -0,0 +1,366 @@
+            pool_id = self.find_pool_for_underlyings(underlyings)
+            pool = self.safe.contract(self.vault.getPool(pool_id)[0])
+        else:
+            pool_id = pool.getPoolId()
+
+        underlyings, reserves, _ = self.vault.getPoolTokens(pool_id)
+        gauge = self.safe.contract(self.gauge_factory.getPoolGauge(pool))
+
+        amount_in = pool.balanceOf(self.safe)
+
+        if unstake:
+            amount_in += gauge.balanceOf(self.safe)
+
+        underlying_index = underlyings.index(asset.address)
+
+        data_encoded = eth_abi.encode_abi(
+            ['uint256', 'uint256', 'uint256'],
+            [0, amount_in, underlying_index]
+            )
+
+
+        if self.pool_type(pool_id) == 'Stable':
+            underlying_out = StableMath.calcTokenOutGivenExactBptIn(
+                pool, reserves, underlying_index, amount_in
+            )
+        else:
+            underlying_out = WeightedMath.calc_token_out_given_exact_bpt_in(
+                pool, reserves[underlying_index], amount_in, underlying_index
+            )
+
+        min_out = underlying_out * (1 - self.max_slippage)
+
+        if is_eth:
+            underlyings = list(underlyings)
+            weth_index = underlyings.index(registry.eth.treasury_tokens.WETH)
+            underlyings[weth_index] = ZERO_ADDRESS
+
+        request = (
+            underlyings,
+            [min_out if x == asset.address else 0 for x in underlyings],
+            data_encoded,
+            False
+        )
+
+        pool.approve(self.vault, amount_in)
+
+        self._unstake_and_withdraw_all(
+            pool, request, unstake, claim, destination
+        )
+
+
+    def _unstake_and_withdraw_all(
+        self, pool, request, unstake, claim, destination
+    ):
+        destination = self.safe if not destination else destination
+        pool_id = pool.getPoolId()
+
+        if unstake:
+            gauge = self.safe.contract(self.gauge_factory.getPoolGauge(pool))
+            balance_before_gauge = gauge.balanceOf(self.safe)
+            gauge.withdraw(gauge.balanceOf(self.safe), claim)
+            assert pool.balanceOf(self.safe) >= balance_before_gauge
+
+        balances_before = [Contract(x).balanceOf(destination) for x in request[0]]
+
+        self.vault.exitPool(
+            pool_id, self.safe, destination, request
+        )
+
+        balances_after = [Contract(x).balanceOf(destination) for x in request[0]]
+        assert any([x > y for x, y in zip(balances_after, balances_before)])
+
+
+    def claim_all(self, underlyings=None, pool=None):
+        # claim reward token from pool's gauge
+        if underlyings:
+            underlyings = sorted([x.address for x in underlyings])
+            pool_id = self.find_pool_for_underlyings(underlyings)
+            pool = self.safe.contract(self.vault.getPool(pool_id)[0])
+
+        gauge = self.safe.contract(self.gauge_factory.getPoolGauge(pool))
+        reward_address = gauge.reward_tokens(0)
```

Comment:
minor comment here, not sure how often that may happen that a gauge has multiple rewards tokens for the grabs, but perhaps worthy to leave it more dynamic this method based on the `gauge.reward_count()`, instead of hardcoding the checkup on only one token

Reply:
good idea